site/content/blog/2020-12-04-common-i18n-mistakes.md

```diff
@@ -0,0 +1,315 @@
+
+Because we support so many languages, we have to be aware of how to keep our applications properly translatable. There's many easy mistakes you can make when writing an application in one language that can make it difficult to translate into others. We use [react-intl](https://formatjs.io/docs/react-intl/) in our client-side applications and [go-i18n](https://github.com/nicksnyder/go-i18n) on the server which both offer a variety of features to help with this, but we can still cause problems if we're not careful when using translated text. Here's a few examples of common problems that we've run into and how to solve them.
+
+## Mistake 1: Not translating something
+
+This one is definitely the easiest to avoid, but it can still catch people who aren't used to writing applications outside their native language.
+
+For example, suppose someone is adding a new button to the web app using React. If they forget to make it to use react-intl, they may write something like this:
+
+```jsx
+function MyButton() {
+    return <button>{'Click me!'}</button>;
+}
+```
+
+That's fine if only English-speaking people use your application, but that doesn't work for us. Instead, we can use react-intl's `FormattedMessage` component to do the translation for us.
+
+```jsx
+import {FormattedMessage} from 'react-intl';
+
+function MyButton() {
+    return (
+        <button>
+            <FormattedMessage
+                id='my_button.label'
+                defaultMessage='Click me!'
+            />
+        </button>
+    );
+}
+```
+
+The code ends up being a bit longer, but that's more than worth it for making the app translatable.
+
+## Mistake 2: Hard coding translations
+
+This mistake is a bit rarer since it requires mixing translated and non-translated text, but it comes from a real-world example in Mattermost.
+
+When you reply to an older message in Mattermost, we display "Commented on Billy's message:" to tell everyone that you're responding to Billy. For a while after we had translation support added, the code for that looked like this:
+
+```jsx
+let username = '@' + rootPost.username;
+if (!username.endsWith('s')) {
+    username += '\'s';
+}
+
+return (
+    <FormattedMessage
+        id='post.commentedOn'
+        defaultMessage='Commented on {username} message:'
+        values={{username}}
+    />
+)
+```
+
+Here, we're using react-intl to add Billy's name into the text as a value, but we're making a mistake here by trying to make the username into a possessive adjective ourselves. By manually adding the "'s", we're applying English grammar regardless of the user's language setting, so this label won't make sense in other languages.
+
+Instead, the "'s" should be part of the translation string. That required changing our grammar rules slightly since we previously avoided adding "'s" to the end of names that ended in the letter S, but the new way is still correct by other popular grammar guidelines. Isn't English great?
```

Comment:
```suggestion
Instead, the "'s" should be part of the translation string. That required changing our grammar rules slightly since we previously avoided adding "'s" to the end of names that ended in the letter "s", but the new way is still correct according to other popular grammar guidelines. Isn't English great?
```